### PR TITLE
FP8 kv cache quantization

### PR DESCRIPTION
**Pull request by CUHKSZzxy**
## Summary

Add PyTorch-backend FP8 KV-cache quantization for paged attention.

The default `fp8` policy uses normal FP8 KV cache with scalar per-tensor K/V scales, matching the common vLLM behavior and avoiding per-token/per-head scale metadata on the hot path.

## What Changed

- Add PyTorch KV-cache support for:
  - `fp8` / `fp8_e4m3`: E4M3 FP8 KV cache
  - `fp8_e5m2`: E5M2 FP8 KV cache
- Store K/V cache as `torch.float8_e4m3fn` or `torch.float8_e5m2`.
- Add scalar `k_scale` / `v_scale` state on attention layers.
- Add FP8 paths for cache fill, paged decode, and prefill flatten.
- Keep normal FP8 metadata-free: no per-token/per-head scale tensors are allocated.
- Add CLI/config parsing and focused kernel/e2e tests.

## Usage

```bash
# no KV-cache quantization
lmdeploy serve api_server <model> --backend pytorch --quant-policy 0

# FP8 E4M3 KV cache
lmdeploy serve api_server <model> --backend pytorch --quant-policy fp8

# FP8 E5M2 KV cache
lmdeploy serve api_server <model> --backend pytorch --quant-policy fp8_e5m2
```

## End-to-End Benchmark

Model: Qwen3.5-35B-A3B  
Backend: LMDeploy PyTorch, TP=2  
Dataset: ShareGPT  
Baseline: BF16 KV, `--quant-policy 0`  
Candidates: FP8 E4M3 KV, `--quant-policy fp8`; FP8 E5M2 KV, `--quant-policy fp8_e5m2`

| Output len | Prompts | BF16 KV tok/s | FP8 E4M3 KV tok/s | E4M3 delta | FP8 E5M2 KV tok/s | E5M2 delta |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| None | 1000 | 5835.55 | 5842.50 | +0.12% | 5617.81 | -3.73% |
| 2048 | 1000 | 15091.14 | 17384.36 | +15.20% | 17158.46 | +13.70% |
| 4096 | 500 | 13480.78 | 15083.34 | +11.89% | 14451.87 | +7.20% |
| 8192 | 200 | 7199.99 | 7390.87 | +2.65% | 8236.96 | +14.40% |

| Output len | Prompts | BF16 KV TTFT ms | FP8 E4M3 KV TTFT ms | E4M3 delta | FP8 E5M2 KV TTFT ms | E5M2 delta |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| None | 1000 | 3669.65 | 3629.63 | +1.09% | 4914.25 | -33.92% |
| 2048 | 1000 | 3809.86 | 3678.69 | +3.44% | 3866.09 | -1.48% |
| 4096 | 500 | 1925.52 | 2018.82 | -4.85% | 1903.33 | +1.15% |
| 8192 | 200 | 979.81 | 983.78 | -0.41% | 965.03 | +1.51% |

Positive TTFT delta means lower/better latency. Small TTFT deltas should be treated as noise.

## Accuracy Check

Model: Qwen3.5-397B-A17B-FP8  
Backend: LMDeploy PyTorch, TP=8  
Candidate: FP8 KV, `--quant-policy fp8`

| Benchmark | FP8 KV cache quant | Official Qwen3.5-397B-A17B-FP8 | Delta | Note |
| --- | ---: | ---: | ---: | --- |
| AIME2025 repeat-32 | 95.83 | 91.3 | +4.53 | Official card reports AIME26, not AIME2025. |
| GPQA_diamond repeat-4 | 86.99 | 88.4 | -1.41 | Official card reports GPQA; benchmark naming may differ. |

## Validation

- `tests/pytorch/kernel/test_fill_kv_cache.py`
- `tests/pytorch/kernel/test_flatten_kv_cache.py`
- `tests/pytorch/kernel/test_paged_attention.py`
- `tests/test_lmdeploy/test_fp8_kv_cache_policy.py`
- `tests/test_lmdeploy/test_quant_policy.py`
- Qwen3.5-35B-A3B TP=2 ShareGPT e2e benchmark above

## Assistance

Assisted with Codex + GPT-5.5 High

### CODE REVIEW

File: lmdeploy/messages.py [hunk 1]
```diff
@@ -457,8 +464,14 @@ def __post_init__(self):
         assert self.max_prefill_token_num >= 0, \
             'invalid max_prefill_token_num'
         assert self.num_gpu_blocks >= 0, 'invalid num_gpu_blocks'
-        assert self.quant_policy in (QuantPolicy.NONE, QuantPolicy.INT4, QuantPolicy.INT8, QuantPolicy.TURBO_QUANT), \
-               'invalid quant_policy'
+        assert self.quant_policy in (
```
Comment (grimoire) [inline lmdeploy/messages.py]:
```
try:
    self.quant_policy = QuantPolicy(self.quant_policy)
except ValueError as e:
    raise ValueError(f'invalid quant_policy: {self.quant_policy}') from e
```
might be better for the check

File: lmdeploy/pytorch/backends/cuda/attention/default.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ class TritonAttentionMetadata(AttentionMetadata):
         q_seqlens: Length of each query sequence [batch_size].
         kv_start_loc: Start location of each KV sequence [batch_size].
         kv_seqlens: Length of each KV sequence [batch_size].
-        quant_policy: Quantization policy (0=none, 4=int4, 8=int8/fp8).
+        quant_policy: Quantization policy (0=none, 4=int4, 8=int8, 16/17=per-tensor fp8).
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/attention/default.py:27]: Since we would send quant_policy here, can we reuse k_scales_zeros instead of create new variable?

File: lmdeploy/pytorch/kernels/cuda/flatten_kv_cache.py [hunk 3]
```diff
@@ -355,6 +437,43 @@ def flatten_kv_cache(k_caches: Tensor,
             BLOCK_DK=BLOCK_DK,
             BLOCK_DV=BLOCK_DV,
         )
+    elif quant_policy in (QuantPolicy.FP8, QuantPolicy.FP8_E5M2):
+        if k_scale is None:
+            k_scale = torch.ones((), device=k_caches.device, dtype=torch.float32)
```
Comment (grimoire) [inline lmdeploy/pytorch/kernels/cuda/flatten_kv_cache.py]: Use lru_cache to avoid repeat ones kernel, or fuse it into the kernel.

File: lmdeploy/pytorch/engine/cache_engine.py [hunk 4]
```diff
@@ -51,6 +51,35 @@ def _get_kv_cache_dtype(model_config: ModelConfig):
...
+def _get_fp8_cache_dtype(quant_policy: QuantPolicy):
+    """Get the cache tensor dtype for an FP8 KV-cache quant policy."""
+    if quant_policy == QuantPolicy.FP8:
+        return torch.float8_e4m3fn
+    if quant_policy == QuantPolicy.FP8_E5M2:
+        return torch.float8_e5m2
+    raise ValueError(f'Not an FP8 quant policy: {quant_policy}')
+
+
+def _describe_kv_cache_quant_policy(quant_policy: QuantPolicy):
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/cache_engine.py:81]: Could be implemented as a dict.

File: lmdeploy/pytorch/nn/attention.py [hunk 5]
```diff
@@ -68,6 +69,11 @@ def __init__(
             self.alibi_ready = False
         else:
             self.alibi_ready = True
+        scale_device = kwargs.get('device', None)
+        # Regular PyTorch FP8 KV cache intentionally uses fixed scalar scales
+        # of 1.0 for now. Kernels accept scale tensors to stay general.
+        self.register_buffer('k_scale', torch.ones((), dtype=torch.float32, device=scale_device))
```
Comment (grimoire) [inline lmdeploy/pytorch/nn/attention.py:75]: Have you test sleep/wakeup?

**Review by grimoire — APPROVED** [review]
